lib/mix/lib/mix/dep/loader.ex

```diff
@@ -301,17 +304,18 @@ defmodule Mix.Dep.Loader do
     |> elem(0)
   end
 
-  defp rebar_children(app, root_config, extra, manager) do
+  defp rebar_children(root_config, manager) do
     from = Path.absname("rebar.config")
     Mix.Rebar.recur(root_config, fn config ->
-      app
-      |> Mix.Rebar.deps(config, overrides(manager, extra))
-      |> Enum.map(fn dep -> %{to_dep(dep, from, manager) | extra: extra} end)
+      overrides = overrides(manager, config)
```

Comment:
@ericmj The previous version was applying overrides ONLY for sub_dirs and apps and those are not even supported in Rebar 3. This PR changes it and we apply overrides for dependencies exclusively. The idea is to extract the overrides here and store in the dependency as part of the extra. We later take it out from extra and apply it properly.

In other words, the overrides system was not applying to dependencies before. That would probably be a good test to add. Maybe we could add a completely bogus dependency to a rebar project and use `{:overrides, :foo, {:deps, []}}` to completely remove it? Can you write push such a test to this PR. ❤️